### ISSUE TITLE
Consider having a view to login through IAP

### ISSUE THREAD

**Issue by flea89** [potatolondon/djangae#1270]
Currently the IAP login happens in the `AuthenticationMiddleware` and not through a view.
That means we lose the control over when the user is logged in, it just happens transparently.

I think that might cause issues, especially if/when you expect users not to be signed in. On top of that the IAP middleware also creates a user if it doesn't exists.

We should consider having a view where the IAP login/authentication happens instead of hiding it in the middleware. 
@Kazade any thoughts?

* Djangae version: x.x.x / master@commit
* Django version: x.x.x
* App Engine SDK version: 1.9.xx
* Using Cloud SDK: yes/no

**Comment by sleepyjames** [comment]
There is a view, or more specifically a URL, provided see https://github.com/potatolondon/djangae/blob/master/docs/googleauth.md#testing-authentication-locally. This view could be exposed with a named URL though. We should suggest settings [`settings.LOGIN_URL`](https://docs.djangoproject.com/en/3.1/ref/settings/#login-url) to this too.

**Comment by flea89** [comment]
Yeah locally that's the case, It's all handled in the`local_iap_login_middleware`.

However my doubt was in general and in production env.
If you take the "local_iap_login_middleware" for a second, it feels a bit inconsistent that the [AuthenticationMiddleware](https://github.com/potatolondon/djangae/blob/539d787c768f6a6f785f65375fae808ce92147d5/djangae/contrib/googleauth/middleware.py#L95) automatically logs users in if the IAPBackend is installed and the user is not authenticated.

It'd feel more consistent to have the same behaviour as the oauth, where the user is logged in through a view (even if it wouldn't actually require any action from the user) and the middleware duty is to set the user on the request.

TBH we'd probably need to think about the actual use cases and how we you would use IAP in production environment. And at what application layer we want that to be be handled. If we leave it this way the control on when a user is logged in is controlled by you IAP configuration and not by your application code.

ie. 
Given an "internal" Application A with 2 routes:

`/public` available to every employee of an org
`/admin` available to a subset of users

You secure the whole application using IAP, with IAP-secured Web App User` set to "org.com".
On application layer you require authentication and authorization to access the `admin` route while you don't for the public.

With the current structure you would have a user authenticated transparently when visiting `/public` (even a user created in your database if the user is a new one), which is likely something you do not want? or am I missing something?

Possibly there are different angles to look at this though? And I was saying before we need to think a bit more how we would actually use it.

**Comment by sleepyjames** [comment]
Sorry @flea89 I completely mis-understood your point here. Reading again it sounds like: Do we want to implicitly login users when a resource is protected with IAP and a user has been through IAPs authentication flow?

>  it feels a bit inconsistent that the AuthenticationMiddleware automatically logs users in if the IAPBackend is installed and the user is not authenticated

Is this true? For the IAPBackend to authenticate a user successfully they must have been through IAPs auth flow for the [required HTTP headers to be set](https://github.com/potatolondon/djangae/blob/master/djangae/contrib/googleauth/backends/iap.py#L23).

> With the current structure you would have a user authenticated transparently when visiting /public (even a user created in your database if the user is a new one), which is likely something you do not want? or am I missing something?

I think creating a user automatically here is OK in the common case. It matches the behavior of `djangae.contrib.gauth` with the App Engine User API in Djangae 1.x . We're authenticating and not applying any authorization logic (we don't set `is_staff` or `is_superuser`) so mixing public and admin resources is safe as long as the user enforces the appropriate tests on their views/middleware, like `django.contrib.auth.decorators.staff_member_required` for example.

It's also worth noting that mixing public(ish) and protected content via IAP doesn't really fit it's usecase, at least the way I read it. You're [prevented](https://cloud.google.com/iap/docs/managing-access) from applying *any* URL conditions when the `allAuthenticated` user's group is added.
